src/main/java/com/eighthlight/fabrial/http/request/RequestReader.java

```diff
@@ -40,27 +39,21 @@ private static RequestBuilder withRequestLine(String requestLine) throws Request
     }
   }
 
-  public Request readRequest() throws RequestParsingException {
+  public Optional<Request> readRequest() throws RequestParsingException {
     try {
       // must pass the reader through each step, otherwise the first reader
       // will consume the entire stream
       var lineReader = new HttpLineReader(is);
-      return Result
-          .attempt(() -> {
-            var firstLine = lineReader.readLine();
-            return Optional.ofNullable(firstLine)
-                           .orElseThrow(() -> new RequestParsingException("Request is empty"));
-          })
-          .flatMapAttempt(RequestReader::withRequestLine)
-          .flatMapAttempt(b -> {
-            var headerReader = new HttpHeaderReader(is);
-            var headers = headerReader.readHeaders();
-            return b.withHeaders(headers);
-          })
-          .map(b -> {
-            return b.withBody(is).build();
-          })
-          .orElseThrow();
+      var firstLine = lineReader.readLine();
+      if (firstLine == null || firstLine.isEmpty()) {
+        return Optional.empty();
+      }
+      return Optional.of(
+          RequestReader.builderWithRequestLine(firstLine)
+                       .withHeaders(new HttpHeaderReader(is).readHeaders())
+                       .withBody(is)
+                       .build()
+      );
```

Comment:
Cool; this got a lot simpler-looking.

Reply:
Yeah, as much as I like my `Result` class, I need to be wary of using it "just because." 😝